crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -636,17 +632,15 @@ unsafe fn get_component_and_ticks(
     entity: Entity,
     location: EntityLocation,
 ) -> Option<(Ptr<'_>, &UnsafeCell<ComponentTicks>)> {
-    let archetype = &world.archetypes[location.archetype_id];
     let component_info = world.components.get_info_unchecked(component_id);
     match component_info.storage_type() {
         StorageType::Table => {
-            let table = &world.storages.tables[archetype.table_id()];
+            let table = &world.storages.tables[location.table_id];
             let components = table.get_column(component_id)?;
-            let table_row = archetype.entity_table_row(location.index);
             // SAFETY: archetypes only store valid table_rows and the stored component type is T
             Some((
-                components.get_data_unchecked(table_row),
-                components.get_ticks_unchecked(table_row),
+                components.get_data_unchecked(location.table_row as usize),
```

Comment:
Same comments here: we should be propagating the type change, rather than casting.

Reply:
Going to defer this to #4878, since this PR is focused on the `EntityLocation` perf imrpovements.